packages/celltags-extension/src/celltag.tsx

```diff
@@ -0,0 +1,204 @@
+    if (!this._editing) {
+      (event.target as HTMLInputElement).blur();
+    }
+  }
+
+  private _onAddTagBlur(input: HTMLInputElement) {
+    if (this._editing) {
+      this._editing = false;
+      this._emptyAddTag(input);
+    }
+  }
+
+  private _onChange(event: React.ChangeEvent<HTMLInputElement>) {
+    if (!event.target.value) {
+      this._emptyAddTag(event.target);
+    } else {
+      const tmp = document.createElement('span');
+      tmp.className = 'add-tag';
+      tmp.textContent = event.target.value;
+      // set width to the pixel length of the text
+      document.body.appendChild(tmp);
+      event.target.style.width = tmp.getBoundingClientRect().width + 8 + 'px';
+      document.body.removeChild(tmp);
+    }
+  }
+
+  private _onAddTagClick(
+    props: FieldProps,
+    event: React.MouseEvent<HTMLElement>
+  ) {
+    const elem = (event.target as HTMLElement).closest('div');
+    const input = elem?.childNodes[0] as HTMLInputElement;
+    if (!this._editing) {
+      this._editing = true;
+      input.value = '';
+      input.focus();
+    } else if (event.target !== input) {
+      this.addTag(props, input.value);
+    }
+    event.preventDefault();
+  }
+
+  private _onTagClick(props: FieldProps, tag: string) {
+    const data = props.formData;
+    if (data.includes(tag)) {
+      data.splice(data.indexOf(tag), 1);
+    } else {
+      data.push(tag);
+    }
+
+    props.formContext.updateMetadata({ [props.name]: data }, true);
+  }
+
+  render(props: FieldProps): JSX.Element {
+    const allTags: string[] = this.pullTags();
+
+    return (
+      <div className="cell-tags">
+        <div className="jp-FormGroup-fieldLabel jp-FormGroup-contentItem">
+          Cell Tags
+        </div>
+        {allTags &&
+          allTags.map((tag: string, i: number) => (
+            <div
+              key={i}
+              className={`tag ${
+                props.formData.includes(tag) ? 'applied-tag' : 'unapplied-tag'
+              }`}
+              onClick={() => this._onTagClick(props, tag)}
+            >
+              <div className="tag-holder">
+                <span>{tag}</span>
+                {props.formData.includes(tag) && (
+                  <LabIcon.resolveReact
+                    icon={checkIcon}
+                    tag="span"
+                    elementPosition="center"
+                    height="18px"
+                    width="18px"
+                    marginLeft="5px"
+                    marginRight="-3px"
+                  />
+                )}
+              </div>
+            </div>
+          ))}
+        <div className="tag unapplied-tag">
+          <div
+            className="tag-holder"
+            onMouseDown={(e: React.MouseEvent<HTMLElement>) =>
+              this._onAddTagClick(props, e)
+            }
+          >
+            <input
+              className="add-tag"
+              type="text"
+              style={{ width: '49px' }}
```

Comment:
Should either css variable or relative length unit like em be used here to make sure width of the element is responsive (to for example work area or viewport width changes)?

Also, should we opt for class names + css sheets vs inline styling? CC @krassowski

Reply:
Added a `jp-CellTags-Empty` class and replaced the `width: '49px'` by a `width: '4em'` in css sheets.

Reply:
Thank you @brichet